docs: kubernetes joining guide + reference (#20298)
gravitational/teleport: docs/pages/management/guides/joining-services-kubernetes-serviceaccount.mdx

---

> title: Joining Services via Kubernetes ServiceAccount token

Reviewer: ```suggestion
title: Joining Services via Kubernetes ServiceAccount Token
```

> description: Use Kubernetes ServiceAccount tokens to join services running in the same Kubernetes cluster as the Auth service.

Reviewer: ```suggestion
description: Use Kubernetes ServiceAccount tokens to join services running in the same Kubernetes cluster as the Auth Service.
```

---

This guide will explain how to use the **Kubernetes join method** to configure

> Teleport Services to join your Teleport cluster without sharing any

Reviewer: ```suggestion
Teleport services to join your Teleport cluster without sharing any
```
I know this was capitalized in the Core Concepts page at one point, but I've changed this so it doesn't seem like "Teleport Service" is a discrete product/architectural component

secrets when running in the same Kubernetes cluster than the Auth Service.

<Notice type="warning">

> The Kubernetes join method is not available in Teleport Cloud as it requires the

Reviewer: ```suggestion
The Kubernetes join method is not available in Teleport Enterprise Cloud as it requires the
```

joining service to run in the same Kubernetes cluster as the Auth Service.

</Notice>

The Kubernetes join method is available in self-hosted versions of Teleport 12+.
It supports joining any Teleport Service running in the same Kubernetes cluster
as the Auth Service.

## Prerequisites

- A running Teleport cluster in Kubernetes. For details on how to set this up,
  see [Guides for running Teleport using Helm](../../deploy-a-cluster/helm-deployments.mdx).
- Editor access to the Kubernetes cluster running the Teleport cluster.
  You must be able to create Namespaces and Deployments.
- A Teleport user with `access` role, or any other role that allows access to
  applications with the label `app: demo-app`

> - Either the Teleport `editor` role, or the ability to `kubectl exec` into your

Reviewer: ```suggestion
- Either the Teleport `editor` role or the ability to `kubectl exec` into your
```

Create a file called `token.yaml` with the following content, which includes
an `allow` rule specifying the Kubernetes namespace and Kubernetes
ServiceAccount in which your Teleport services are running.

```yaml
# token.yaml
kind: token
version: v2
metadata:
  # The token name is not a secret as the Kubernetes join method relies on the
  # Kubernetes signature to establish trust and not on the join token name.
  name: kubernetes-token
  # set a long expiry time, the default for tokens is only 30 minutes
  expires: "2050-01-01T00:00:00Z"
spec:
  # Use the minimal set of system roles required.
  # Common roles are:
  # - "Node" for SSH Service
  # - "Proxy" for Proxy Service
  # - "Kube" for Kube Service
  # - "App" for Application Access Service
  # - "Db" for Database Access Service
  # - "WindowsDesktop" for Windows Desktop Service
  # - "Discovery" for Discovery Service
  roles: [App]

  # set the join method allowed for this token
  join_method: kubernetes

  kubernetes:
    allow:
      # Service account names follow the format "namespace:serviceaccountname".
      - service_account: "teleport-agent:teleport-app-service"
```

> Kubernetes join tokens can be used by any non-Auth Teleport service:
> Kubernetes, Application Access, Database Access, Windows Desktop, ... In this
> guide, we restrict the token to Application Service joining.

Reviewer: ```suggestion
Kubernetes join tokens can be used by any Teleport service besides the Auth Service, such as the Proxy Service and SSH Service. In this
guide, we restrict the token to joining an Application Service instance.
```

It is not recommended to use a single token that can join everything. You should

> restrict the token to the roles used by the joining instance. e.g. a Teleport
> instance running both the App Service and DB Service should use a token with
> `roles: [App, Db]`.

Reviewer: ```suggestion
restrict the token to the roles used by the joining instance. For example, a Teleport
instance running both the Application Service and Database Service should use a token with
`roles: [App, Db]`.
```
For consistency with the rest of the docs

> 

Reviewer: ```suggestion
Follow the instructions below to create the token depending on whether you have administrative access to the Auth Service pod:
```

<Tabs>
<TabItem label="Using tctl locally as your current user">
Make sure your local `tctl` is at least at version 12 with `tctl version`.

> Run `tctl create token.yaml` to create the token.

Reviewer: ````suggestion
Create the token:

```code
$ tctl create token.yaml
```
````
In case users want to copy/paste

```

Finally, validate the token was successfully created:

```code
$ kubectl exec -i -n teleport deployment/teleport-auth tctl get token/kubernetes-token

kind: token
metadata:
  expires: "3000-01-01T00:00:00Z"
  name: kubernetes-token
spec:
  join_method: kubernetes
  roles:
  - App
version: v2
```

</TabItem>
</Tabs>

## Step 2/5. Deploy a demonstration HTTP app

In this step, we deploy a demonstration HTTP application and don't expose it
publicly. Instead, we will manage access to this application with the Teleport
Application Service, which we will register with Teleport using the Kubernetes
join method.

```code
$ kubectl create namespace demo-app
namespace/demo-app created

$ kubectl create deployment --image=nginx --namespace demo-app --port=80 demo-app
deployment.apps/demo-app created

$ kubectl expose deployment demo-app -n demo-app --port=80 --target-port=80 --selector='app=demo-app'
service/demo-app exposed
```

Validate the application pods are running and ready with the following command:

```code
$ kubectl get pods -n demo-app
NAME                        READY   STATUS    RESTARTS   AGE
demo-app-7664d59cb8-bvbmz   1/1     Running   0          67s
```

> ## Step 3/5. Configure your Application Access Service

Reviewer: ```suggestion
## Step 3/5. Configure the Application Service
```

Configure the `teleport-kube-agent` chart to deploy Teleport instances running

> the Application Access Service by creating a `values.yaml` file with the

Reviewer: ```suggestion
the Application Service by creating a `values.yaml` file with the
```

following content:

```yaml
# values.yaml

# Public address of the Teleport cluster with port.
# You must replace the placeholder with your proxy address.
proxyAddr: "teleport.example.com:443"

# Comma-separated list of services the `teleport-kube-agent` chart must run
# (supported values are: kube,db,app)
# In this guide we only deploy app access.
# Adding more services here also requires to add role to the provision token created in step 1.
roles: app

joinParams:
  method: "kubernetes"
  # this must match the provision token created in Step 1.
  tokenName: "kubernetes-token"
  
apps:
  - name: demo-app
    uri: "http://demo-app.demo-app.svc.cluster.local:80"
```

> ## Step 4/5. Deploy the Application Access Service

Reviewer: ```suggestion
## Step 4/5. Deploy the Application Service
```

To use the token created in Step 1, the joining instance must run in the same
Kubernetes cluster as the Auth Service and have a Kubernetes ServiceAccount
token mounted.

> [The `teleport-kube-agent` chart](../../reference/helm-reference/teleport-kube-agent.mdx)
> will take care of this by default.

Reviewer: ```suggestion
The `teleport-kube-agent` chart that you will install in this section will take care of this by default.
```
I don't think we need to link the user to another docs page since we are referring to this chart in the current step.

> Deploy the Teleport Application Access Service by running the following command:

Reviewer: ```suggestion
Deploy the Teleport Application Service by running the following command:
```
